README.md
<img src="https://fontpreview.readthedocs.io/en/latest/_static/fp.png" alt="fontpreview package">
<br>

[![Codacy Badge](https://app.codacy.com/project/badge/Grade/a4e407dc1c5747c08fe8b1d984cb2676)](https://www.codacy.com/gh/MatteoGuadrini/fontpreview/dashboard?utm_source=github.com&amp;utm_medium=referral&amp;utm_content=MatteoGuadrini/fontpreview&amp;utm_campaign=Badge_Grade)
[![CircleCI](https://circleci.com/gh/MatteoGuadrini/fontpreview.svg?style=svg)](https://circleci.com/gh/MatteoGuadrini/fontpreview)

# fontpreview: Python library for font previews

This is a library that allows you to create preview images from one or more selected fonts.

Full docs is here: [ReadTheDocs](https://fontpreview.readthedocs.io/en/latest/)

## Test

If you want to test the package before installing, use the test:
```console
$ git clone https://github.com/MatteoGuadrini/fontpreview.git
$ cd fontpreview
$ python -m unittest test_fp.py
```

## Installation

Use Pypi:
```console
$ pip install --user fontpreview
```

> **Note**: If you want to use the command line tool, you need to install the system-wide library: `pip install fontpreview`

## Simple usage

Preview example:
```python
from fontpreview import FontPreview

fp = FontPreview('/tmp/noto.ttf')
fp.save('/tmp/fp.png')
```
<img src="https://i.ibb.co/258dCPZ/fp.png" alt="FontPreview object" width="350" height="150">
<br><br>

Banner example:
```python
from fontpreview import FontBanner

fb = FontBanner('/tmp/noto.ttf', 'landscape', bg_color=(153, 153, 255), mode='fontname')
fb.save('/tmp/fb.png')
```
<img src="https://i.ibb.co/FVWdkYC/fb.png" alt="FontBanner object" width="350" height="130">
<br><br>

Logo example:
```python
from fontpreview import FontLogo

fl = FontLogo('/tmp/noto.ttf', 'Fp')
fl.save('/tmp/fl.png')
```
<img src="https://i.ibb.co/j302Y5k/fl.png" alt="FontLogo object">
<br><br>

Font wall example:
```python
from fontpreview import FontBanner, FontWall

# Define the various parts of wall
fb = FontBanner('/tmp/noto.ttf', 'landscape' , mode='fontname')
fb2 = FontBanner('/tmp/noto.ttf', 'landscape' , mode='alpha')
fb3 = FontBanner('/tmp/noto.ttf', 'landscape' , mode='letter')
fb4 = FontBanner('/tmp/noto.ttf', 'landscape' , mode='paragraph')
fw = FontWall([fb,fb2,fb3,fb4])
fw.save('/tmp/fw.png')
```
<img src="https://i.ibb.co/cDBST2r/fw.png" alt="FontWall object" width="650" height="200">
<br><br>


Font page example:
```python
from fontpreview import FontPage, FontBanner

# Define the various parts of page
header = FontBanner('/tmp/noto.ttf', 'landscape' , mode='fontname')
body = FontBanner('/tmp/noto.ttf', 'landscape' , mode='paragraph')
footer = FontBanner('/tmp/noto.ttf', 'landscape' , mode='letter')
# Create FontPage object
fpage = FontPage()
fpage.set_header(header)
fpage.set_body(body)
fpage.set_footer(footer)
# Design all parts
fpage.draw()
fpage.save('/tmp/fpage.png')

```
<img src="https://i.ibb.co/LgFLnXk/fpage.png" alt="FontPage object" width="650" height="850">
<br><br>

Font page with template example:
```python
from fontpreview import FontPage, FontPageTemplate, FontBanner

# Define the various parts of page
header = FontBanner('/tmp/noto.ttf', 'landscape' , mode='fontname')
body = FontBanner('/tmp/noto.ttf', 'landscape' , mode='paragraph')
footer = FontBanner('/tmp/noto.ttf', 'landscape' , mode='letter')
# Create font page template
template = FontPageTemplate(3508)
template.set_body(170, 1, 'lcenter')
template.set_footer(100, 4, 'lcenter')
# Create FontPage object
fpage = FontPage(template=template)
fpage.set_header(header)
fpage.set_body(body)
fpage.set_footer(footer)
# Design all parts
fpage.draw()
fpage.save('/tmp/fpage_template.png')

```
<img src="https://i.ibb.co/qJjMGpr/fpage-template.png" alt="FontPage with template object" width="650" height="850">
<br><br>

Font booklet example:
```python
from fontpreview import FontPage, FontBanner, FontBooklet

# Define the various parts of page
header = FontBanner('/tmp/noto.ttf', 'landscape' , mode='fontname')
body = FontBanner('/tmp/noto.ttf', 'landscape' , mode='paragraph')
footer = FontBanner('/tmp/noto.ttf', 'landscape' , mode='letter')
# Create FontPage object
fpage1 = FontPage(header=header, body=body, footer=footer)
fpage2 = FontPage(header=header, body=body, footer=footer)
# Design all parts
fpage1.draw()
fpage2.draw()
# Create book
book = FontBooklet(fpage1, fpage2)
book.save('/tmp/noto_book/')        # save page1.png, page2.png in /tmp/noto_book/ folder

```

## Advanced usage

Below is an example of various previews of the _"Fira Code regular"_ font. 
Does it remind you of anything? [Fira code original](https://github.com/tonsky/FiraCode/raw/master/extras/logo.svg)
```python
# FIRA CODE WALL
from fontpreview import FontBanner, FontWall
fira_code = '/tmp/firacode.ttf'
# RGB group = ('background', 'FIRA COD color', 'Ligature color', 'E color background')
colors_group = [
                ('black', (0, 143, 0), (0, 236, 236), (255, 0, 255)),
                ('black', (166, 47, 123), (81, 208, 93), (11, 179, 248)),
                ((13, 21, 43), (112, 204, 84), (226, 110, 34), (223, 245, 90)),
                ((43, 6, 42), (136, 126, 135), (4, 150, 153), (147, 103, 145)),
                ((39, 57, 85), (255, 241, 208), (208, 84, 0), (209, 215, 227)),
                ((31, 63, 89), (248, 248, 242), (230, 219, 117), (166, 226, 51)),
                ((1, 47, 80), (224, 202, 52), (73, 217, 38), (255, 125, 158)),
                ((0, 0, 170), (75, 224, 245), (255, 255, 85), (0, 170, 170)),
                ('white', 'black', 'black', 'black'),
                ((247, 247, 247), (167, 29, 93), (121, 93, 163), (0, 134, 179)),
                ((239, 240, 243), (15, 131, 207), (208, 84, 0), (105, 40, 122)),
                ((239, 231, 212), (218, 116, 53), (0, 142, 212), (186, 136, 0)),
                ((39, 40, 34), (132, 214, 45), (249, 39, 114), (174, 129, 255)),
                ((43, 48, 59), (180, 142, 173), (143, 161, 179), (152, 190, 140)),
                ((32, 32, 32), (171, 130, 84), (160, 171, 127), (216, 127, 98)),
                ((0, 43, 54), (0, 160, 153), (126, 143, 3), (218, 66, 130))
                ]
banners = []
# Create banners
for colors in colors_group:
    # Create a FontBanner objects
    fb = FontBanner(fira_code, (413, 240))
    liga = FontBanner(fira_code, (413, 240))
    E = FontBanner(fira_code, (40, 70))
    # Set background colors
    fb.bg_color = liga.bg_color = colors[0]
    E.bg_color = colors[3]
    # Set foreground colors
    fb.fg_color = colors[1]
    liga.fg_color = colors[2]
    E.fg_color = colors[0]
    # Set text
    fb.font_text = 'FIRA COD'
    liga.font_text = "!=->>++:="
    E.font_text = 'E'
    # Set text position
    E.set_text_position('ltop')
    fb.set_text_position((25, 60))
    liga.set_text_position('top')
    # Adding image on fb
    fb.add_image(liga, (0, 122))
    fb.add_image(E, (339, 60))
    # Add to list of banners
    banners.append(fb)

# Create a wall
fw = FontWall(banners, max_tile=4)
fw.save('/tmp/fira_code.png')
```
<img src="https://i.ibb.co/cvnPRdB/fira-code.png" alt="Fira code wall">
<br><br>

## Command line interface
Along with the package, a command line tool based on this python package is installed. 
The class used for command line previews is FontPreview. For all the options of this tool, 
see the [docs](https://fontpreview.readthedocs.io/en/latest/cli.html), otherwise run `fp -h`.

```console
$ fp /tmp/noto.ttf
```

This command save a *fontpreview.png* in a current directory.

<img src="https://i.ibb.co/258dCPZ/fp.png" alt="FontPreview object" width="350" height="150">
<br><br>

## Open source
_fontpreview_ is a open source project. Any contribute, It's welcome.

**A great thanks**.

For donations, press this

For me

[![paypal](https://www.paypalobjects.com/en_US/i/btn/btn_donateCC_LG.gif)](https://www.paypal.me/guos)

For [Telethon](http://www.telethon.it/)

The Telethon Foundation is a non-profit organization recognized by the Ministry of University and Scientific and Technological Research.
They were born in 1990 to respond to the appeal of patients suffering from rare diseases.
Come today, we are organized to dare to listen to them and answers, every day of the year.

<a href="https://www.telethon.it/sostienici/dona-ora"> <img src="https://www.telethon.it/dev/_nuxt/img/c6d474e.svg" alt="Telethon" title="Telethon" width="200" height="104" /> </a>

[Adopt the future](https://www.ioadottoilfuturo.it/)


## Acknowledgments

Thanks to Mark Lutz for writing the _Learning Python_ and _Programming Python_ books that make up my python foundation.

Thanks to Kenneth Reitz and Tanya Schlusser for writing the _The Hitchhiker’s Guide to Python_ books.

Thanks to Dane Hillard for writing the _Practices of the Python Pro_ books.

Special thanks go to my wife, who understood the hours of absence for this development. 
Thanks to my children, for the daily inspiration they give me and to make me realize, that life must be simple.

Thanks Python!


CHANGES.md
# Release notes

## 1.2.1
Jul 18, 2023

- Fixed text centering offset in *FontPreview* class

## 1.2.0
Dec 12, 2020

- Additions show method on *FontPreview*, *FontWall*, *FontPage* class
- Additions _FontBooklet_ class
- Fix FontPageTemplate *\__str\__* method
- Fix message raises on set_header/logo/body/footer method on _FontPage_ class

## 1.1.0
Dec 12, 2020

- Additions declaration mode for *FontPage* class
- Additions *test_hex_color* method on TestFontPreview class
- Additions *test_hex_color* method on TestFontPreview class
- Additions *test_declarative_object* method on TestFontPreview class
- Additions *test_other_color_system* method on TestFontPreview class
- Fix TestFontPreview.test_font_size method and FontPage.__compose method
- Fix test_text_position method on TestFontPreview class

## 1.0.0
Dec 5, 2020

- *fontpreview* python package is available.
- *fp* command line tool is available.
- *FontPreview* class represents a preview object of a given font.
- *FontBanner* class represents a preview banner object of a given font. Based on *FontPreview* class.
- *FontLogo* class represents a logo object of a given font. Based on *FontPreview* class.
- *FontWall* class represents a wall of preview fonts object.
- *FontPage* class represents a page of preview fonts object.
- *FontPageTemplate* class represents a template for *FontPage* class.
- Documentation has been created


test_fp.py
import unittest
import os
from fontpreview import FontPreview, FontBanner, FontLogo, FontWall, FontPage, FontPageTemplate, FontBooklet

# Enter the file font to test and check exists
font = input('Enter path of font file to test: ')
if not os.path.exists(font):
    raise OSError("file font {0} doesn't exists".format(font))


class TestFontPreview(unittest.TestCase):
    fp, fb, fl = FontPreview(font), FontBanner(font), FontLogo(font, 'Fl')
    fw, fpage, fpage_t = FontWall([fb]), FontPage(), FontPageTemplate(3508)
    book = FontBooklet(fpage, fpage)

    def test_instance(self):
        # test if instance has been created
        self.assertIsInstance(self.fp, FontPreview)
        self.assertIsInstance(self.fb, FontBanner)
        self.assertIsInstance(self.fl, FontLogo)
        self.assertIsInstance(self.fw, FontWall)
        self.assertIsInstance(self.fpage, FontPage)
        self.assertIsInstance(self.fpage_t, FontPageTemplate)
        self.assertIsInstance(self.book, FontBooklet)

    def test_set_color_with_name(self):
        # change background color
        self.fp.bg_color = self.fb.bg_color = self.fl.bg_color = 'blue'
        # change background color
        self.fp.fg_color = self.fb.fg_color = self.fl.fg_color = 'yellow'
        # test draw it
        self.fp.draw()
        self.fb.draw()
        self.fl.draw()

    def test_set_color_with_tuple(self):
        # change background color
        self.fp.bg_color = self.fb.bg_color = self.fl.bg_color = (51, 153, 193)
        # change background color
        self.fp.fg_color = self.fb.fg_color = self.fl.fg_color = (253, 194, 45)
        # test draw it
        self.fp.draw()
        self.fb.draw()
        self.fl.draw()

    def test_set_dimension(self):
        # test dimension of FontPreview object
        self.fp.dimension = (1000, 1000)
        # test draw it
        self.fp.draw()
        self.assertEqual(self.fp.image.size, self.fp.dimension)
        # test dimension of FontBanner object
        self.fb.set_orientation((1000, 1000))
        # test draw it
        self.fb.draw()
        self.assertEqual(self.fb.image.size, self.fb.dimension)
        # test dimension of FontBanner object and font position
        self.fb.set_orientation('landscape', 'lcenter')
        # test draw it
        self.fb.draw()
        self.assertEqual(self.fb.image.size, self.fb.dimension)
        # test dimension of FontLogo object
        self.fl.new_size((75, 75))
        self.fl.new_size((100, 100))
        self.fl.new_size((150, 150))
        self.fl.new_size((170, 170))
        # test draw it
        self.fp.draw()
        self.assertEqual(self.fl.image.size, self.fl.dimension)
        # test dimension of FontPage
        self.fpage.set_header(self.fb)
        self.fpage.set_body(self.fb)
        self.fpage.set_footer(self.fb)
        self.fpage.dimension = (1000, 1000)
        # test draw it
        self.fpage.draw()
        self.assertEqual(self.fpage.page.size, self.fpage.dimension)

    def test_add_image(self):
        # test add image to FontBanner
        self.fb.add_image(self.fp, (500, 500))
        # test resize in add image
        fb_big = FontBanner(font, (2000, 2000))
        self.fb.add_image(fb_big, (500, 500))

    def test_font_size(self):
        # Test FontPreview font size
        self.fp.set_font_size(70)
        self.assertEqual(self.fp.font.size, 70)
        # Test FontBanner font size
        self.fb.set_font_size(80)
        self.assertEqual(self.fb.font.size, 80)
        # Test FontBanner font size
        self.fl.set_font_size(50)
        self.assertEqual(self.fl.font.size, 50)
        # Test FontPage and FontPageTemplate font size
        template = FontPageTemplate()
        template.set_header(90, 1, 'lcenter')
        template.set_body(90, 3, 'lcenter')
        template.set_footer(90, 2, 'lcenter')
        fpage = FontPage(template=template)
        fpage.set_header(self.fb)
        fpage.set_body(self.fb)
        fpage.set_footer(self.fb)
        fpage.draw()
        # FontPageTemplate font size
        self.assertEqual(template.header_font_size, 90)
        self.assertEqual(template.body_font_size, 90)
        self.assertEqual(template.footer_font_size, 90)
        # FontPage font size
        self.assertEqual(fpage.header.font.size, 90)
        self.assertEqual(fpage.body.font.size, 90)
        self.assertEqual(fpage.footer.font.size, 90)

    def test_text_position(self):
        # Test FontPreview font size
        self.fp.set_text_position('lcenter')
        # Test FontBanner font size
        self.fb.set_text_position('rcenter')
        # Test FontBanner font size
        self.fl.set_text_position('center')
        # Test FontPreview font size
        self.fp.set_text_position((100, 100))
        # Test FontBanner font size
        self.fb.set_text_position((100, 100))
        # Test FontBanner font size
        self.fl.set_text_position((100, 100))

    def test_set_text(self):
        # Test FontPreview font size
        self.fp.font_text = 'unittest'
        self.fp.draw(align='center')
        self.assertEqual(self.fp.font_text, 'unittest')
        # Test FontBanner font size
        self.fb.set_mode('fontname')
        self.fb.draw(align='center')
        self.assertEqual(self.fb.font_text, '{0}'.format(self.fb.font.getname()[0]))
        # Test FontBanner font size
        self.fl.font_text = 'ut'
        self.fl.draw(align='center')
        self.assertEqual(self.fl.font_text, 'ut')

    def test_add_to_wall(self):
        # Create FontWall
        fw = FontWall([self.fb, self.fp, self.fl])
        self.assertIsInstance(fw, FontWall)
        fw.draw(fw.max_tile)
        # Create FontWall with max_tile args
        fw = FontWall([self.fb, self.fp, self.fl], max_tile=3)
        self.assertIsInstance(fw, FontWall)
        fw.draw(fw.max_tile)
        # Create FontWall with max_tile args and
        fw = FontWall([self.fb, self.fp, self.fl], max_tile=3, mode='vertical')
        self.assertIsInstance(fw, FontWall)
        fw.draw(fw.max_tile)

    def test_template_page(self):
        # Create FontPage
        page = FontPage(template=self.fpage_t)
        # Test instance
        self.assertIsInstance(page, FontPage)
        # Test method FontPageTemplate
        self.fpage_t.set_header(120, 1, 'center')
        self.fpage_t.set_body(170, 3, 'left')
        self.fpage_t.set_footer(100, 2, 'right')
        # Test value of header
        self.assertEqual(self.fpage_t.header_text_position, 'center')
        self.assertEqual(self.fpage_t.header_font_size, 120)
        self.assertEqual(self.fpage_t.header_units, self.fpage_t.unit * 1)
        # Test value of header
        self.assertEqual(self.fpage_t.body_text_position, 'left')
        self.assertEqual(self.fpage_t.body_font_size, 170)
        self.assertEqual(self.fpage_t.body_units, self.fpage_t.unit * 3)
        # Test value of header
        self.assertEqual(self.fpage_t.footer_text_position, 'right')
        self.assertEqual(self.fpage_t.footer_font_size, 100)
        self.assertEqual(self.fpage_t.footer_units, self.fpage_t.unit * 2)

    def test_declarative_object(self):
        # FontPreview object
        fp = FontPreview(font,
                         font_size=50,
                         font_text='some text',
                         color_system='RGB',
                         bg_color='blue',
                         fg_color='yellow',
                         dimension=(800, 400))
        # FontBanner object
        fb = FontBanner(font,
                        orientation='portrait',
                        bg_color='blue',
                        fg_color='yellow',
                        mode='paragraph',
                        font_size=70,
                        color_system='RGB')
        # FontLogo object
        fl = FontLogo(font,
                      'Fl',
                      size=(170, 170),
                      bg_color='yellow',
                      fg_color='blue',
                      font_size=50,
                      color_system='RGB')
        # FontPage object
        fpage = FontPage(header=fb, logo=fl, body=fb, footer=fb)
        # test if instance has been created
        self.assertIsInstance(fp, FontPreview)
        self.assertIsInstance(fb, FontBanner)
        self.assertIsInstance(fl, FontLogo)
        self.assertIsInstance(fpage, FontPage)

    def test_other_color_system(self):
        fp = self.fp
        fb = self.fb
        fl = self.fl
        # define new color system
        fp.color_system = 'CMYK'
        # change background color
        fp.bg_color = fb.bg_color = fl.bg_color = (51, 153, 193)
        # change background color
        fp.fg_color = fb.fg_color = fl.fg_color = (253, 194, 45)
        # test draw it
        fp.draw()
        fb.draw()
        fl.draw()

    def test_hex_color(self):
        fp = self.fp
        fb = self.fb
        fl = self.fl
        # define new color system
        fp.color_system = 'CMYK'
        # change background color
        fp.bg_color = fb.bg_color = fl.bg_color = "#269cc3"
        # change background color
        fp.fg_color = fb.fg_color = fl.fg_color = "#ff0000"
        # test draw it
        fp.draw()
        fb.draw()
        fl.draw()

    def test_parse_not_fontpage_on_fontbooklet(self):
        self.assertRaises(ValueError, FontBooklet, self, self.fp, self.fw)

    def test_iter_fontbooklet(self):
        self.fpage.set_header(self.fb)
        self.fpage.set_body(self.fb)
        self.fpage.set_footer(self.fb)
        for page in self.book:
            page.draw()


if __name__ == '__main__':
    unittest.main()


fontpreview/__init__.py
#!/usr/bin/env python3
# -*- encoding: utf-8 -*-
# vim: se ts=4 et syn=python:

# created by: matteo.guadrini
# __init__.py -- fontpreview
#
#     Copyright (C) 2020 Matteo Guadrini <matteo.guadrini@hotmail.it>
#
#     This program is free software: you can redistribute it and/or modify
#     it under the terms of the GNU General Public License as published by
#     the Free Software Foundation, either version 3 of the License, or
#     (at your option) any later version.
#
#     This program is distributed in the hope that it will be useful,
#     but WITHOUT ANY WARRANTY; without even the implied warranty of
#     MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
#     GNU General Public License for more details.
#
#     You should have received a copy of the GNU General Public License
#     along with this program.  If not, see <http://www.gnu.org/licenses/>.
from pkg_resources import get_distribution, DistributionNotFound

from .fontpreview import FontPreview
from .fontbanner import FontBanner, FontLogo, FontWall
from .fontpage import FontPage, FontPageTemplate, FontBooklet

try:
    __version__ = get_distribution('fontpreview').version
except DistributionNotFound:
    __version__ = 'version not found'


fontpreview/__version__.py
#!/usr/bin/env python3
# -*- encoding: utf-8 -*-
# vim: se ts=4 et syn=python:

# created by: matteo.guadrini
# __version__ -- fontpreview
#
#     Copyright (C) 2020 Matteo Guadrini <matteo.guadrini@hotmail.it>
#
#     This program is free software: you can redistribute it and/or modify
#     it under the terms of the GNU General Public License as published by
#     the Free Software Foundation, either version 3 of the License, or
#     (at your option) any later version.
#
#     This program is distributed in the hope that it will be useful,
#     but WITHOUT ANY WARRANTY; without even the implied warranty of
#     MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
#     GNU General Public License for more details.
#
#     You should have received a copy of the GNU General Public License
#     along with this program.  If not, see <http://www.gnu.org/licenses/>.
VERSION = '0.9.0'

fontpreview/fontbanner.py
#!/usr/bin/env python3
# -*- encoding: utf-8 -*-
# vim: se ts=4 et syn=python:

# created by: matteo.guadrini
# fontbanner -- fontpreview
#
#     Copyright (C) 2020 Matteo Guadrini <matteo.guadrini@hotmail.it>
#
#     This program is free software: you can redistribute it and/or modify
#     it under the terms of the GNU General Public License as published by
#     the Free Software Foundation, either version 3 of the License, or
#     (at your option) any later version.
#
#     This program is distributed in the hope that it will be useful,
#     but WITHOUT ANY WARRANTY; without even the implied warranty of
#     MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
#     GNU General Public License for more details.
#
#     You should have received a copy of the GNU General Public License
#     along with this program.  If not, see <http://www.gnu.org/licenses/>.

# region imports
import os
from .fontpreview import FontPreview
from PIL import Image


# endregion

# region functions
def resize(image, bg_image):
    """
    Resize image

    :param image: image to resize
    :param bg_image: background image
    :return: Image object
    """
    # Check size of background image
    new_size = image.size
    while new_size > bg_image.size:
        width, height = new_size
        new_size = (int(width // 1.2), int(height // 1.2))
    # Resize image
    return image.resize(new_size)


# endregion

# region classes
class FontBanner(FontPreview):
    """
    Class that represents the banner of a font
    """

    def __init__(self, font,
                 orientation='landscape',
                 bg_color='white',
                 fg_color='black',
                 mode='letter',
                 font_size=64,
                 color_system='RGB'
                 ):
        """
        Object that represents the banner of a font

        :param font: font file
        :param orientation: the orientation of the banner; 'landscape', 'portrait' or tuple(x,y).
        :param bg_color: background color of preview. Default is 'white'.
        :param fg_color: foreground or font color of preview. Default is 'black'.
        :param mode: the text inside the banner; 'letter','fontname', 'paragraph', 'alpha' and 'combination'.
        :param font_size: font size. Default is 64.
        :param color_system: color system string. Default is 'RGB'.
        """
        # Define properties
        FontPreview.__init__(self, font=font,
                             bg_color=bg_color,
                             fg_color=fg_color,
                             font_size=font_size,
                             color_system=color_system)
        self.set_orientation(orientation)
        self.mode = mode
        self.set_text_position('center')
        # Create default image
        self.set_mode(mode=self.mode)

    def __str__(self):
        """
        String representation of font banner

        :return: string
        """
        return FontPreview.__str__(self) + ",mode={mode}".format(mode=self.mode)

    def set_orientation(self, orientation, font_position='center'):
        """
        Set orientation of banner

        :param orientation: the orientation of the banner; 'landscape' or 'portrait'
        :param font_position: font position respect dimension of banner
        :return: None
        """
        # Calculate banner size
        if isinstance(orientation, tuple):
            self.dimension = orientation
            # Recalculate font position
            self.set_text_position(font_position)
            return None
        else:
            LANDSCAPE = (1653, 560)
            PORTRAIT = (560, 1653)
            if orientation == 'landscape':
                self.dimension = LANDSCAPE
            elif orientation == 'portrait':
                self.dimension = PORTRAIT
            else:
                raise ValueError('orientation is "landscape","portrait" or tuple(x,y)')
            # Recalculate font position
            self.set_text_position(font_position)

    def set_mode(self, mode, align='center'):
        """
        Set the text mode

        :param mode: mode that sets the text in the banner
        :param align: alignment of text. Available 'left', 'center' and 'right'
        :return: None
        """
        MODE = {
            'letter': 'a b c d e f\ng h i j k l\nm n o p q r\ns t u v w x y z',
            'alpha': 'Aa Bb Cc Dd Ee Ff\n1 2 3 4 5 6 7 8 9 0',
            'fontname': '{0}'.format(self.font.getname()[0]),
            'paragraph': 'Lorem ipsum dolor sit amet,\nconsectetur adipiscing elit.',
            'combination': '{0}\n{1}'.format(self.font.getname(),
                                             'Lorem ipsum dolor sit amet,\nconsectetur adipiscing elit.'
                                             ),
            'none': ''
        }
        # Verify is mode exists
        if mode in MODE:
            self.mode = mode
            self.font_text = MODE.get(mode)
            # Create default image
            self.draw(align=align)
        else:
            raise ValueError('mode is "letter", "alpha", "fontname", "paragraph" and "combination"')

    def add_image(self, image, position):
        """
        Adds an additional image to the banner

        :param image: path of image
        :param position: position of image
        :return: None
        """
        # Create image
        if isinstance(image, FontPreview):
            img = image.image
        else:
            img = Image.open(image)
        # Check if the image is bigger than the banner
        if img.size > self.dimension:
            img = resize(img, self.image)
        # Add image
        self.image.paste(img, position)


class FontLogo(FontPreview):
    """
    Class that represents the logo of a font
    """

    def __init__(self, font,
                 letters,
                 size=(100, 100),
                 bg_color='white',
                 fg_color='black',
                 font_size=64,
                 color_system='RGB'
                 ):
        """
        Object that represents the logo of a font

        :param font: font file
        :param letters: One or two letters (or anything)
        :param size: size of logo square. Default is (100, 100)
        :param bg_color: background color of preview. Default is 'white'.
        :param fg_color: foreground or font color of preview. Default is 'black'.
        :param font_size: font size. Default is 64.
        :param color_system: color system string. Default is 'RGB'.
        """
        FontPreview.__init__(self, font=font,
                             bg_color=bg_color,
                             fg_color=fg_color,
                             font_size=font_size,
                             color_system=color_system)
        # Check if the letters exceed the number 2
        if len(letters) > 2:
            raise ValueError('letters can be maximum two')
        else:
            self.font_text = letters
        # Check maximum size
        self.__max_size(size)
        # Set letter position
        self.set_text_position('center')
        # Built a logo font
        self.draw()

    def __max_size(self, size):
        """
        Check maximum size

        :param size: New size
        :return: None
        """
        max_size = ((75, 75), (100, 100), (150, 150), (170, 170))
        if size in max_size:
            self.dimension = size
        else:
            raise ValueError('The max size of the logo can be this: (75, 75), (100, 100), (150, 150), (170, 170)')

    def new_size(self, size):
        """
        Define new size of FontLogo object

        :param size: size of fontlogo object
        :return: None
        """
        # Check maximum size
        self.__max_size(size)
        # Built a logo font
        self.set_text_position('center')


class FontWall:
    """
    Class that represents the wall of fonts
    """

    def __init__(self, fonts, max_tile=2, mode='horizontal'):
        """
        Object that represents the wall of fonts

        :param fonts: font list; string or FontPreview object
        :param max_tile: maximum tile per row/column
        :param mode: image alignment, 'horizontal' or 'vertical'
        """
        # Check if list contains string or FontPreview object
        if isinstance(fonts, list):
            self.fonts = []
            for font in fonts:
                if isinstance(font, FontPreview):
                    self.fonts.append(font)
                else:
                    _font = FontBanner(font)
                    self.fonts.append(_font)
        else:
            raise TypeError("'fonts' must be a list")
        # Other properties
        self.color_system = 'RGB'
        self.bg_color = 'white'
        self.max_width = None
        self.max_height = None
        self.mode = mode
        self.max_tile = max_tile
        # Build the wall
        self.wall = None
        self.draw(self.max_tile)

    def __str__(self):
        """
        String representation of font wall

        :return: string
        """
        return str(["tile{0}={1}".format(i, f) for i, f in enumerate(self.fonts)])

    def __concatenate(self, fonts, position):
        """
        Link multiple images to form a layout inside the wall

        :param fonts: list of FontPreview
        :param position: paste positions
        :return: tuple
        """
        # Get max width and height, presume horizontal
        if self.mode == 'horizontal':
            max_width = sum([font.image.width for font in fonts])
            max_height = max([font.image.height for font in fonts])
        else:
            max_width = max([font.image.width for font in fonts])
            max_height = sum([font.image.height for font in fonts])
        # Create background
        dst = Image.new(self.color_system, (max_width, max_height))
        start = 0
        for font in fonts:
            # Compose the row
            if self.mode == 'horizontal':
                dst.paste(font.image, (start, 0))
                start = font.image.width + start
            # Compose the column
            elif self.mode == 'vertical':
                dst.paste(font.image, (0, start))
                start = font.image.height + start
            else:
                raise ValueError("the mode can be 'horizontal' or 'vertical'")
        self.wall.paste(dst, position)
        return dst.size

    def draw(self, max_tile):
        """
        Draw wall with fonts on properties of this object

        :param max_tile: maximum tile per row
        :return: None
        """
        # Split fonts into maximum tile per row/column
        fonts = []
        for i in range(0, len(self.fonts), max_tile):
            fonts.append(self.fonts[i:i + max_tile])
        # Calculate max_width and max_height of wall
        max_width = []
        max_height = []
        for font in fonts:
            if self.mode == 'horizontal':
                max_width.append(sum([f.image.width for f in font]))
                max_height.append(max([f.image.height for f in font]))
            else:
                max_width.append(max([f.image.width for f in font]))
                max_height.append(sum([f.image.height for f in font]))
        if self.mode == 'horizontal':
            self.max_width = max(max_width)
            self.max_height = sum(max_height)
        else:
            self.max_width = sum(max_width)
            self.max_height = max(max_height)
        self.wall = Image.new(self.color_system, (self.max_width, self.max_height), color=self.bg_color)
        # Build the wall
        start_position = (0, 0)
        for font in fonts:
            if self.mode == 'horizontal':
                last_position = self.__concatenate(font, start_position)
                start_position = (0, (start_position[1] + last_position[1]))
            else:
                last_position = self.__concatenate(font, start_position)
                start_position = ((start_position[0] + last_position[0]), 0)

    def save(self, path=os.path.join(os.path.abspath(os.getcwd()), 'fontwall.png')):
        """
        Save the font wall

        :param path: path where you want to save the font wall
        :return: None
        """
        self.wall.save(path)

    def show(self):
        """
        Displays this image.

        :return: None
        """
        self.wall.show()

# endregion
